source/common/buffer/buffer_impl.cc

```diff
@@ -218,32 +218,28 @@ void* OwnedImpl::linearize(uint32_t size) {
   if (slices_.empty()) {
     return nullptr;
   }
-  uint64_t linearized_size = 0;
-  uint64_t num_slices_to_linearize = 0;
-  for (const auto& slice : slices_) {
-    num_slices_to_linearize++;
-    linearized_size += slice->dataSize();
-    if (linearized_size >= size) {
-      break;
-    }
-  }
-  if (num_slices_to_linearize > 1) {
-    auto new_slice = OwnedSlice::create(linearized_size);
+  if (slices_[0]->dataSize() < size) {
+    auto new_slice = OwnedSlice::create(size);
     uint64_t bytes_copied = 0;
-    Slice::Reservation reservation = new_slice->reserve(linearized_size);
+    Slice::Reservation reservation = new_slice->reserve(size);
```

Comment:
One functional difference between the old and new is that `size` is reserved, rather than total data size of the buffer's slices up to `size` -- based on usage in the description, it seems like this isn't an issue since linearize is meant to produce fixed size buffers?

Reply:
size contains the total size of the data that will be copied.  The original implementation had a local variable linearized_size that used to hold the sum of the data slices that would be fully copied to the linearized region.  The invariant used to be size <= linearized_size.

New implementation only copies 'size' bytes, so we only need to reserve that much.